design/bios-config.md

```diff
@@ -0,0 +1,151 @@
+      * [Summary](#summary)
+      * [Motivation](#motivation)
+         * [Goals](#goals)
+         * [Non-Goals](#non-goals)
+      * [Proposal](#proposal)
+         * [Implementation Details/Notes/Constraints [optional]](#implementation-detailsnotesconstraints-optional)
+         * [Risks and Mitigations](#risks-and-mitigations)
+      * [Design Details](#design-details)
+         * [Work Items](#work-items)
+         * [Dependencies](#dependencies)
+         * [Test Plan](#test-plan)
+         * [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+         * [Version Skew Strategy](#version-skew-strategy)
+      * [Drawbacks [optional]](#drawbacks-optional)
+      * [Alternatives [optional]](#alternatives-optional)
+      * [References](#references)
+
+<!-- Added by: stack, at: 2019-02-15T11:41-05:00 -->
+
+<!--te-->
+
+## Summary
+
+This document explains the usage of YAML attributes for BIOS configuration for the BareMetalHost of the BMO. These configurations will be applied to the host according to the vendor's BMC driver being used by Ironic.
+
+## Motivation
+
+We need this document to specify the typical usage of BIOS YAML attributes in the BareMetalHost YAML. 
+
+
+### Goals
+
+1. To agree on the format of the attributes that are to be supported as a starting point
+2. To keep the attribute naming vendor agnostic
+
+### Non-Goals
+
+1. To list every attribute for every vendor exhaustively
+
+## Proposal
+The proposal is to set/unset the BIOS configuration values using vendor drivers available for each BMC type. For this purpose, new YAML attributes are to be introduced in the BareMetalHost spec. 
+ 
+### Implementation Details/Notes/Constraints 
+All the BIOS related attributes or fields will come under the sub-section called ```bios``` in spec. The values for each attribute can be a boolean or a string. The vendor driver type to implement this configuration is not to be specified separetely as it is already known from the bmc sub-section.
+Initially, the BIOS configuration can be applied using the controller types idrac and irmc
+
+```yaml
+apiVersion: metal3.io/v1alpha1
+kind: BareMetalHost
+metadata:
+ name: bm0
+spec:
+ online: true
+ bmc:
+   address: idrac://192.168.122.1:6230/
+   credentialsName: bm0-bmc-secret
+ bootMACAddress: 52:54:00:b7:b2:6f	
+ bios:
+   EnableSriovGlobal: false
+   BootMode: uefi
+   EnableVirtualization: false
+   EnableHyperThreading: true
+   EnableAdjCacheLine: false
+```
+
+The user can only specify the BIOS values related to the BMC type being used as they will be validated accordingly.
+
+One gotcha to this is about hiding the vendor sprawl in the parameter names and coming up with generic titles for the config params. This would then need to be handled in the operator code (the vendor-specific implementation for Ironic API calls). Meaning, it will still be required to map these *generic* names to the actual parameter names in the API call for it to function.
```

Comment:
Regarding vendor specific calls. We have a bios setting interface with-in ironic to raise visibility and be used across vendors, the issue at hand is one of the major vendors has not implemented a mapping to the new interface and only has what is referred to as a "vendor passthru" interface for bios settings which takes a separate path and is different API calls along with an entirely different interaction which is outside of cleaning and clean steps. While a wrapper should be fairly simple for that driver with-in ironic, it seems like a hidden detail that is not accounted for if there is not willingness to work with the ironic community to get the necessary bios interface code in place for the driver in question.

Reply:
Can you recall what vendor is this? It'll help with the investigation

Reply:
The vendor in question which only supports bios settings through their "vendor passthru" API interface is the same vendor you have focused this pull request on. I'm hoping to drive forward adoption of the standard interface in the coming months, but community assistance would help tremendously.

Reply:
This is an interesting fact I wasn't aware of. Can you give me some references for my own study? I'd like to study this in detail. And regarding the work on adoption of the standard interface, I'm willing to assist in that regard.

Reply:
Our concern primarily, is to come up with consistent naming conventions for the BIOS params common across all vendors (as a starting step). The API calls and implementation details are a separate concern.

Regarding vendor-passthru, yes you're right, it follows a separate set of API calls and implementation. But what we're proposing is the idrac ```vendor-interface``` invokation for BIOS configs. Further details [here](https://docs.openstack.org/ironic/latest/admin/drivers/idrac.html). We have successfully tested this interface to apply BIOS configs in clean steps of ironic. 

About the standard ```bios-interface```, I agree that it currently doesn't support BIOS configs with ```idrac hardware type``` and I'm willing to contribute in that regard.

Reply:
We've decided to pursue the vendor passthru path to configure BIOS for iDRAC in the operator. Later on, down the road, when the ```bios_interface``` has been implemented with iDRAC driver, this passthru implementation will be deprecated and later removed, in favor of that standard ``bios_interface``` implementation. I think this should seal the deal here.

Reply:
> We've decided to pursue the vendor passthru path to configure BIOS for iDRAC in the operator.

Does that imply a different set of calls to ironic, or just different arguments to the existing calls?

Reply:
A different set of API calls

Reply:
> A different set of API calls

So then the provisioner needs some way to know how to make the right set of calls. What interface will provide that information?

Reply:
Yes it will need to know that. If you mean in Ironic, then it's the ```vendor-interface```. The passthru methods are to be implemented in the ironic provisioner of BMO for this to be accomplished.

Reply:
> I believe since the passthru methods are not typical API calls, they are to be implemented in the ironic provisioner of BMO differently. I'm unsure if any interface can _provide_ these API calls to us.

By "interface", I meant which type within the metal3 code would have the knowledge of which ironic calls to make differently based on the hardware type.

So far we've only used the [BMC AccessDetails interface](https://github.com/metal3-io/baremetal-operator/blob/master/pkg/bmc/access.go#L27) to pass different data to the same calls. Now we're talking about making completely different calls to ironic based on the hardware. I'm not sure I'm entirely comfortable doing that, so I'm trying to understand how you intend to do it and what impact that will have on the code.

Does it mean we need an entirely different [Provisioner](https://github.com/metal3-io/baremetal-operator/blob/master/pkg/provisioner/provisioner.go#L27) implementation?

Does the AccessDetails interface need to have new methods to give hints to the existing Ironic provisioner? An AccessDetails implementation doesn't have to know anything about Ironic right now. How much would we want them to have to know?

Do we need to define a new interface to pass to the existing Provisioner? Would that be useful with other Provisioner implementations?

Reply:
I should also ask, what would be involved in making it possible to use the same set of API calls in Ironic for all hardware types?

Reply:
I see the access details interface has ```VendorInterface``` in it already. The API calls in Ironic that are needed for this configuration use the ```VendorInterface``` to be performed.

> Does it mean we need an entirely different Provisioner implementation?

Also, I don't think there will be a need for an *entirely different provisioner* implementation.

> Does the AccessDetails interface need to have new methods to give hints to the existing Ironic provisioner? An AccessDetails implementation doesn't have to know anything about Ironic right now. How much would we want them to have to know?

No, I don't think it will be necessary to add new methods to it. Also, it's best to not have that interface know anything about ironic at all. I agree with that. And we won't need to expose any ironic details to BMO.

> Do we need to define a new interface to pass to the existing Provisioner? Would that be useful with other Provisioner implementations?

I don't think it necessary as of now. And regarding other provisioniner implementation, I'm not very sure about that, I guess it will depend on what other provisioner we're talking about.

Reply:
> I should also ask, what would be involved in making it possible to use the same set of API calls in Ironic for all hardware types?

That's a great question actually. I'm in the process of creating a complete flow of this and will share very soon. It'll help us figure out what exactly will drive the implementation - in terms of function calls and code impacts.

Reply:
> > I should also ask, what would be involved in making it possible to use the same set of API calls in Ironic for all hardware types?
> 
> That's a great question actually. I'm in the process of creating a complete flow of this and will share very soon. It'll help us figure out what exactly will drive the implementation - in terms of function calls and code impacts.

Great, it will be very useful to understand those differences.

Reply:
Going back on some stuff and reading more, here's my proposal for ```ironic.go``` module.

First, we make a private function ```buildTargetBIOSConfig(biosConfig *metal3v1alpha1.BIOSConfig)```, this will generate a structure we need to pass into the passthru call using the YAML parameters from the CRD.

Then we create a performBIOSConfig() function, which make the API calls to configure BIOS.

These can be placed within the startProvisioning() function which is then called with Provision() function of the provisioner interface. Within the ```manageable``` state case.

The ```vendor-interface``` with ```idrac``` hardware type will be invoked to issue these API calls.

Reply:
No, they are two separate mechanisms for interacting with the ```Node``` object in Ironic.

The cleaning method is what is used either _before_ ```deploying``` a ```Node``` (with a ```clean``` API call when ```Node``` is in ```manageable``` state) or _after_ ```deleting``` a ```Node``` (with a ```clean``` API call when ```Node``` is in the ```deleted``` state.

This interaction is a prioritized sequence of work items ```clean steps```. These employ the standard interfaces like ```power```, ```boot```, ```raid``` and ```bios``` to _clean_ the ```Node``` by issuing required API calls through those interfaces to get the work done. You typically do this like so
1. Bring the ```Node``` into ```manageable``` state.
2. Create the ```target``` JSON with a steps you need to perform, say powerOn->Boot->BIOS->RAID.
3. Issue the ```clean``` API call with ```Node``` UUID and the ```target``` as the payload (either manually or through Ironic CLI)
4. Receive the JSON and extract required data from it. Or just a status if operation wasn't supposed to return any data.

The ```passthru``` mechanism is a bit different and has a limited/specific scope. It is also invoked from the ```manageable``` state of ```Node``` but rather discretely. It comprises API calls that **_directly_** invoke the ```vendor-interface``` of the ```Node``` and call some ```passthru method```s that provide some specific functionality.

For Dell's iDRAC ```hardware type```, the BIOS configuration is not supported through the standard ```bios-interface``` of ```Node``` yet, but a set of ```passthru method``` calls. Therefore, the interaction is outside the scope of ```clean steps```. And the method you shared above doesn't apply in this case.

In summary, the ```passthru``` mechanism for us would work something like this.
1. Bring the ```Node``` into ```manageable``` state.
2. Issue the ```passthru``` Ironic API call on the ```Node``` with a payload that identifies the ```passthru method``` we're using, for instance ```get_bios_config```.
3. Receive the JSON and extract the required data from it.
4. Perform error handling as necessary with HTTP status codes.

Reply:
> For Dell's iDRAC `hardware type`, the BIOS configuration is not supported through the standard `bios-interface` of `Node` yet, but a set of `passthru method` calls. Therefore, the interaction is outside the scope of `clean steps`. And the method you shared above doesn't apply in this case.
> 
> In summary, the `passthru` mechanism for us would work something like this.
> 
> 1. Bring the `Node` into `manageable` state.
> 2. Issue the `passthru` Ironic API call on the `Node` with a payload that identifies the `passthru method` we're using, for instance `get_bios_config`.
> 3. Receive the JSON and extract the required data from it.
> 4. Perform error handling as necessary with HTTP status codes.

The whole point of using Ironic is to isolate these hardware-specific steps at that layer. I am not enthusiastic about having to include hardware-specific processes in metal3 as well.

What would it take to make the iDrac driver in Ironic support BIOS configuration via cleaning?

Reply:
I totally agree that it's not ideal, but that's the way it is right now, unfortunately for us.

It would require implementation of the ```bios-interface``` for the iDRAC driver, with the ```wsman``` API calls necessary to perform the required BIOS configurations. That's a lot of work on its own. Which **_is_** indeed the long term plan for the iDRAC driver, but for the time being, we have no other choice but to stick to the ```vendor-interface``` ```passthru method```s.

Since the code impact is minimal, we can comment these specific methods heavily to ensure they are identified within the ```ironic.go``` module. I'd really love to get started on the work Doug.

Reply:
> For Dell's iDRAC `hardware type`, the BIOS configuration is not supported through the standard `bios-interface` of `Node` yet

Sooo it's hard to read this and not think that it sounds like a problem to be solved in Ironic. Like, Ironic's whole job is to abstract between different drivers, we shouldn't be having to do that again in Metal³, especially not when we're already trying to abstract over Ironic. If some hardware vendor is unable for whatever reason to offer some capabilities then maybe that's an issue Mr. Invisible Hand can help out with.

That said though, realistically we can acknowledge that the `pkg/bmc` package is already ironic-specific (maybe it should be `pkg/provisioner/ironic/bmc`). The stuff that `DriverInfo()` returns is a straight input to the Ironic API. So adding more provisioner-specific methods to that interface doesn't actually change the relationship between parts of our code.

Reply:
> Sooo it's hard to read this and not think that it sounds like a problem to be solved in Ironic. Like, Ironic's whole job is to abstract between different drivers, we shouldn't be having to do that again in Metal³, especially not when we're already trying to abstract over Ironic. If some hardware vendor is unable for whatever reason to offer some capabilities then maybe that's an issue Mr. Invisible Hand can help out with.

I agree. Wait, who's Mr Invisible Hand?

 
> That said though, realistically we can acknowledge that the `pkg/bmc` package is already ironic-specific (maybe it should be `pkg/provisioner/ironic/bmc`). The stuff that `DriverInfo()` returns is a straight input to the Ironic API. So adding more provisioner-specific methods to that interface doesn't actually change the relationship between parts of our code.

That's true. That's why I believe we're ready to start the work on this.

Reply:
> It would require implementation of the bios-interface for the iDRAC driver, with the wsman API calls necessary to perform the required BIOS configurations. That's a lot of work on its own.

It may or may not be a lot of work, but it has to be done nonetheless.

> Which is indeed the long term plan for the iDRAC driver, but for the time being, we have no other choice but to stick to the vendor-interface passthru methods.

Just to make something clear to everyone: vendor passthru calls have much weaker guarantees than other APIs. They are by definition unstable (no backward compatibility) and bypass most sanity checks (so may harm if implemented or used incorrectly). The former means that we may remove them (with appropriate deprecation, of course) when the BIOS interface approach is finally implemented.

Reply:
I agree. It must be done sooner than later.

Regarding the unstability, of course, upon implementing the standard ```bios-interface``` the passthru will be deprecated but will be alongside a patch to the provisioner for implementing BIOS with clean steps.
And I can ensure the implementation will be tested thoroughly with most stringent sanity conditions. I think we should move forward.

Reply:
> I agree. It must be done sooner than later.
> 
> Regarding the unstability, of course, upon implementing the standard `bios-interface` the passthru will be deprecated but will be alongside a patch to the provisioner for implementing BIOS with clean steps.
> And I can ensure the implementation will be tested thoroughly with most stringent sanity conditions. I think we should move forward.

I understand that you want to move ahead with this work, but I am not prepared to accept an implementation that relies on the vendor pass-through APIs in Ironic. I don't want separate code paths in the provisioner here in metal3, since we already have a layer for isolating vendor differences in Ironic. We seem to all agree that also having vendor-specific code paths here isn't ideal, and I see no particular reason to rush down a path none of us wants to take. I'm sure we can work with the Ironic team to give timely reviews to the required changes there.

Reply:
I see. So, other than this implementation detail, is there anything else in the proposal that we think is undecided? 

We must be settled on everything else before I move ahead with ironic contributions.

Reply:
> > It would require implementation of the bios-interface for the iDRAC driver, with the wsman API calls necessary to perform the required BIOS configurations. That's a lot of work on its own.
> 
> It may or may not be a lot of work, but it has to be done nonetheless.
> 
Less than you'd think. It looks like the wsman vendor pass-through based interface and it looks like it takes a key-value in the form of setting-value. So potentially, they could surface a bios interface that uses the same backend library calls that already exist in their python-dracclient library. On the pure redfish side, it is a lot of work because of the job creation process that is required.

Reply:
I was also thinking the same, when looking at the code later. A lot of framework is already in place for this. Let me see what I can do.
Besides, you'd already expressed on the same discussion, earlier that you're looking for community contribution on this, so I guess it's a win-win for everybody?

Reply:
Given that Dell is switching away from WSMAN, an investment in a non-WSMAN approach is urgently needed. The current vendor passthru is WSMAN-only, which makes me even less inclined to use it.

I'm not sure a job creation is a lot of work, to be honest, they already do a similar trick for virtual media. It just will be a custom `idrac-redfish` BIOS interface.